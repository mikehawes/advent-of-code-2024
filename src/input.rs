use std::fs::read_to_string;
use std::io;
use std::path::{Path, PathBuf};

pub fn input_to_string<P: AsRef<Path>>(path: P) -> io::Result<String> {
    let full_path = file_in_src(path);
    read_to_string(full_path)
}

pub fn file_in_src<P: AsRef<Path>>(path: P) -> PathBuf {
    let mut full_path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    full_path.push("src");
    full_path.push(path);
    full_path
}
