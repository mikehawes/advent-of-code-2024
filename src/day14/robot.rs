pub struct Robot {
    position: Position,
    velocity: Velocity,
}

pub type Position = [usize; 2];

pub type Velocity = [isize; 2];

pub type FloorSize = [usize; 2];

impl Robot {
    fn parse(_: &str) -> Robot {
        Robot {
            position: [0, 0],
            velocity: [0, 0],
        }
    }
    pub fn parse_vec(string: &str) -> Vec<Robot> {
        string.lines().map(Robot::parse).collect()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::input::input_to_string;
    use insta::assert_snapshot;
    use std::collections::HashMap;

    #[test]
    fn can_print_robots() {
        let string = input_to_string("day14/example.txt").unwrap();
        let robots = Robot::parse_vec(&string);
        assert_snapshot!(print(&robots, [11, 7]))
    }

    fn print(robots: &Vec<Robot>, floor: FloorSize) -> String {
        let mut position_to_count: HashMap<Position, usize> = HashMap::new();
        for robot in robots {
            position_to_count
                .entry(robot.position)
                .and_modify(|count| *count += 1)
                .or_insert(1);
        }
        let mut str = String::new();
        for y in 0..floor[1] {
            for x in 0..floor[0] {
                str.push_str(
                    &position_to_count
                        .get(&[x, y])
                        .map(|c| c.to_string())
                        .unwrap_or(".".to_string()),
                );
            }
            str.push('\n');
        }
        str
    }
}
