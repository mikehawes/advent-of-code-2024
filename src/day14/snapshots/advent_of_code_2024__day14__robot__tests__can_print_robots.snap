---
source: src/day14/robot.rs
assertion_line: 36
expression: "print(&robots, [11, 7])"
snapshot_kind: text
---
12..........
...........
...........
...........
...........
...........
...........
